exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/dml/MergeTableTest.kt

```diff
@@ -0,0 +1,489 @@
+                Dest.mergeFrom(Source, on = { Source.key eq Dest.key }) {
+                    updateWhenMatched(deleteWhere = (Dest.value greater 1)) {
+                        it[Dest.value] = Source.value
+                    }
+                }
+            }
+        }
+    }
+
+    object SourceNoAutoId : IdTable<Int>("test_source_no_auto_id") {
+        override val id = integer("id").entityId()
+        val value = varchar("value", 128)
+        override val primaryKey = PrimaryKey(id)
+    }
+
+    object DestNoAutoId : IdTable<Int>("test_dest_no_auto_id") {
+        override val id = integer("id").entityId()
+        val value = varchar("value", 128)
+        override val primaryKey = PrimaryKey(id)
+    }
+
+    @Test
+    fun testAutoGeneratedOnCondition() {
+        withTables(excludeSettings = defaultExcludeSettings, SourceNoAutoId, DestNoAutoId) {
+            SourceNoAutoId.insert {
+                it[id] = 1
+                it[value] = "1"
+            }
+            SourceNoAutoId.insert {
+                it[id] = 2
+                it[value] = "2"
+            }
+            DestNoAutoId.insert {
+                it[id] = 1
+                it[value] = "1"
+            }
+
+            DestNoAutoId.mergeFrom(SourceNoAutoId) {
+                insertWhenNotMatched {
+                    it[DestNoAutoId.id] = SourceNoAutoId.id
+                    it[DestNoAutoId.value] = SourceNoAutoId.value
+                }
+                updateWhenMatched {
+                    it[DestNoAutoId.value] = DestNoAutoId.value + SourceNoAutoId.value
+                }
+            }
+
+            assertEquals("11", DestNoAutoId.selectAll().where { DestNoAutoId.id eq 1 }.single()[DestNoAutoId.value])
+            assertEquals("2", DestNoAutoId.selectAll().where { DestNoAutoId.id eq 2 }.single()[DestNoAutoId.value])
+        }
+    }
+
+    @Test
+    fun testAutoGeneratedOnConditionWithTableAlias() {
+        withTables(excludeSettings = defaultExcludeSettings, SourceNoAutoId, DestNoAutoId) {
+            SourceNoAutoId.insert {
+                it[id] = 1
+                it[value] = "1"
+            }
+            SourceNoAutoId.insert {
+                it[id] = 2
+                it[value] = "2"
+            }
+            DestNoAutoId.insert {
+                it[id] = 1
+                it[value] = "1"
+            }
+
+            val sourceAlias = SourceNoAutoId.alias("test_source_alias")
+            val destAlias = DestNoAutoId.alias("test_dest_alias")
+
+            destAlias.mergeFrom(sourceAlias) {
+                insertWhenNotMatched {
+                    it[destAlias[DestNoAutoId.id]] = sourceAlias[SourceNoAutoId.id]
+                    it[destAlias[DestNoAutoId.value]] = sourceAlias[SourceNoAutoId.value]
+                }
+                updateWhenMatched {
+                    it[destAlias[DestNoAutoId.value]] = destAlias[DestNoAutoId.value] + sourceAlias[SourceNoAutoId.value]
+                }
+            }
```

Comment:
Unless I missed it, there's a lot of tests for single or multiple-of-the-same-type branches, but none with mixed except this one. Could a delete branch be added here so at least one test could have all 3 together?

This is also a good example of a test that could be consolidated with the 1 above.